File: engine/grid/loe.py
"""Line of effect helpers."""

from __future__ import annotations

from typing import Literal

from engine.core.state import BattleState, UnitState
from engine.grid.areas import line_points
from engine.grid.map import in_bounds, is_blocked

CoverGrade = Literal["none", "standard", "greater", "blocked"]


def _sign(value: int) -> int:
    if value > 0:
        return 1
    if value < 0:
        return -1
    return 0


def has_tile_line_of_effect(
    state: BattleState,
    source_x: int,
    source_y: int,
    target_x: int,
    target_y: int,
) -> bool:
    if not in_bounds(state, source_x, source_y):
        return False
    if not in_bounds(state, target_x, target_y):
        return False

    path = line_points(source_x, source_y, target_x, target_y)
    for idx, (x, y) in enumerate(path):
        if idx == 0:
            continue

        prev_x, prev_y = path[idx - 1]
        step_x = x - prev_x
        step_y = y - prev_y
        if abs(step_x) == 1 and abs(step_y) == 1:
            side_a = (prev_x + step_x, prev_y)
            side_b = (prev_x, prev_y + step_y)
            side_a_blocked = in_bounds(state, side_a[0], side_a[1]) and is_blocked(state, side_a[0], side_a[1])
            side_b_blocked = in_bounds(state, side_b[0], side_b[1]) and is_blocked(state, side_b[0], side_b[1])
            if side_a_blocked and side_b_blocked:
                return False

        if idx == len(path) - 1:
            # Allow targeting an occupied endpoint tile.
            return not is_blocked(state, x, y)
        if is_blocked(state, x, y):
            return False
    return True


def cover_grade_between_tiles(
    state: BattleState,
    source_x: int,
    source_y: int,
    target_x: int,
    target_y: int,
) -> CoverGrade:
    if not has_tile_line_of_effect(state, source_x, source_y, target_x, target_y):
        return "blocked"

    sx = _sign(source_x - target_x)
    sy = _sign(source_y - target_y)
    if sx == 0 and sy == 0:
        return "none"

    candidates: list[tuple[int, int]]
    if sx == 0:
        candidates = [(target_x - 1, target_y), (target_x + 1, target_y)]
    elif sy == 0:
        candidates = [(target_x, target_y - 1), (target_x, target_y + 1)]
    else:
        candidates = [(target_x + sx, target_y), (target_x, target_y + sy)]

    blocked_count = 0
    for x, y in candidates:
        if in_bounds(state, x, y) and is_blocked(state, x, y):
            blocked_count += 1

    if blocked_count >= 2:
        return "greater"
    if blocked_count == 1:
        return "standard"
    return "none"


def cover_ac_bonus_from_grade(grade: CoverGrade) -> int:
    if grade == "standard":
        return 2
    if grade == "greater":
        return 4
    return 0


def cover_ac_bonus_between_tiles(
    state: BattleState,
    source_x: int,
    source_y: int,
    target_x: int,
    target_y: int,
) -> int:
    return cover_ac_bonus_from_grade(
        cover_grade_between_tiles(
            state=state,
            source_x=source_x,
            source_y=source_y,
            target_x=target_x,
            target_y=target_y,
        )
    )


def has_line_of_effect(state: BattleState, source: UnitState, target: UnitState) -> bool:
    if not source.alive or not target.alive:
        return False
    return has_tile_line_of_effect(state, source.x, source.y, target.x, target.y)


def cover_grade_for_units(state: BattleState, source: UnitState, target: UnitState) -> CoverGrade:
    if not source.alive or not target.alive:
        return "blocked"
    return cover_grade_between_tiles(state, source.x, source.y, target.x, target.y)


def cover_ac_bonus_for_units(state: BattleState, source: UnitState, target: UnitState) -> int:
    if not source.alive or not target.alive:
        return 0
    return cover_ac_bonus_between_tiles(state, source.x, source.y, target.x, target.y)

File: src/grid/loe.ts
/**
 * Line of effect helpers.
 * Mirrors engine/grid/loe.py
 *
 * Pathfinder 2e ORC cover rules: standard cover (+2 AC), greater cover (+4 AC).
 */

import { BattleState, UnitState, unitAlive } from "../engine/state";
import { linePoints } from "./areas";
import { inBounds, isBlocked } from "./map";

export type CoverGrade = "none" | "standard" | "greater" | "blocked";

function sign(value: number): number {
  if (value > 0) return 1;
  if (value < 0) return -1;
  return 0;
}

export function hasTileLineOfEffect(
  state: BattleState,
  sourceX: number,
  sourceY: number,
  targetX: number,
  targetY: number,
): boolean {
  if (!inBounds(state, sourceX, sourceY)) return false;
  if (!inBounds(state, targetX, targetY)) return false;

  const path = linePoints(sourceX, sourceY, targetX, targetY);
  for (let idx = 0; idx < path.length; idx++) {
    const [x, y] = path[idx];
    if (idx === 0) continue;

    const [prevX, prevY] = path[idx - 1];
    const stepX = x - prevX;
    const stepY = y - prevY;

    // Corner pinch check for diagonal movement
    if (Math.abs(stepX) === 1 && Math.abs(stepY) === 1) {
      const sideA: [number, number] = [prevX + stepX, prevY];
      const sideB: [number, number] = [prevX, prevY + stepY];
      const sideABlocked =
        inBounds(state, sideA[0], sideA[1]) &&
        isBlocked(state, sideA[0], sideA[1]);
      const sideBBlocked =
        inBounds(state, sideB[0], sideB[1]) &&
        isBlocked(state, sideB[0], sideB[1]);
      if (sideABlocked && sideBBlocked) return false;
    }

    if (idx === path.length - 1) {
      // Allow targeting an occupied endpoint tile
      return !isBlocked(state, x, y);
    }
    if (isBlocked(state, x, y)) return false;
  }
  return true;
}

export function coverGradeBetweenTiles(
  state: BattleState,
  sourceX: number,
  sourceY: number,
  targetX: number,
  targetY: number,
): CoverGrade {
  if (!hasTileLineOfEffect(state, sourceX, sourceY, targetX, targetY)) {
    return "blocked";
  }

  const sx = sign(sourceX - targetX);
  const sy = sign(sourceY - targetY);
  if (sx === 0 && sy === 0) return "none";

  let candidates: Array<[number, number]>;
  if (sx === 0) {
    candidates = [
      [targetX - 1, targetY],
      [targetX + 1, targetY],
    ];
  } else if (sy === 0) {
    candidates = [
      [targetX, targetY - 1],
      [targetX, targetY + 1],
    ];
  } else {
    candidates = [
      [targetX + sx, targetY],
      [targetX, targetY + sy],
    ];
  }

  let blockedCount = 0;
  for (const [x, y] of candidates) {
    if (inBounds(state, x, y) && isBlocked(state, x, y)) {
      blockedCount++;
    }
  }

  if (blockedCount >= 2) return "greater";
  if (blockedCount === 1) return "standard";
  return "none";
}

export function coverAcBonusFromGrade(grade: CoverGrade): number {
  if (grade === "standard") return 2;
  if (grade === "greater") return 4;
  return 0;
}

export function coverAcBonusBetweenTiles(
  state: BattleState,
  sourceX: number,
  sourceY: number,
  targetX: number,
  targetY: number,
): number {
  return coverAcBonusFromGrade(
    coverGradeBetweenTiles(state, sourceX, sourceY, targetX, targetY),
  );
}

export function hasLineOfEffect(
  state: BattleState,
  source: UnitState,
  target: UnitState,
): boolean {
  if (!unitAlive(source) || !unitAlive(target)) return false;
  return hasTileLineOfEffect(state, source.x, source.y, target.x, target.y);
}

export function coverGradeForUnits(
  state: BattleState,
  source: UnitState,
  target: UnitState,
): CoverGrade {
  if (!unitAlive(source) || !unitAlive(target)) return "blocked";
  return coverGradeBetweenTiles(state, source.x, source.y, target.x, target.y);
}

export function coverAcBonusForUnits(
  state: BattleState,
  source: UnitState,
  target: UnitState,
): number {
  if (!unitAlive(source) || !unitAlive(target)) return 0;
  return coverAcBonusBetweenTiles(state, source.x, source.y, target.x, target.y);
}
